Refactor credential issuance to use scope-based approach

Awambeng opened the pull request:
This PR refactors the credential issuance logic to adopt a scope-based approach, replacing the previous client-based method. Key changes include:

- Transitioned credential retrieval to utilize the ClientScopeModel for credentials.
- Refined claim assignment logic for better role mapping and type safety.
- Enhanced error handling and scope validation for credential and role claims.
- Updated tests to align with the new scope-based approach.

Closes https://github.com/keycloak/keycloak/issues/32957

Awambeng committed:
Refactor credential issuance to use scope-based approach instead of client-based

- Updated retrieval process to use  for handling credentials.
- Refined claim assignment for improved compatibility with client role mappings.
- Enhanced error handling and logging for scope and role retrieval.
- Updated tests to reflect the scope-based approach for credential issuance.

Closes #32957

Signed-off-by: Awambeng Rodrick <awambengrodrick@gmail.com>

Awambeng committed:
Merge branch 'main' into issue-32957

mposolda commented:
@francis-pouatcha @wistefan @IngridPuppet @tnorimat @thomasdarimont @Captain-P-Goldfish Anyone available to review the PR please?

Captain-P-Goldfish commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/mappers/OID4VCTargetRoleMapper.java:
```diff
@@ -113,31 +114,42 @@ public void setClaimsForSubject(Map<String, Object> claims,
+        ClientRoleModel clientRoleModel = new ClientRoleModel(clientModel.getClientId(), clientRoles);
         Role rolesClaim = toRolesClaim(clientRoleModel);
         if (rolesClaim.getNames().isEmpty()) {
+            LOGGER.debugf("No role names found for client %s, skipping claim assignment.", client);
             return;
         }
-        var modelMap = OBJECT_MAPPER.convertValue(toRolesClaim(clientRoleModel), Map.class);
 
-        if (claims.containsKey(propertyName)) {
-            if (claims.get(propertyName) instanceof Set rolesProperty) {
+        Map<String, Object> modelMap = OBJECT_MAPPER.convertValue(rolesClaim, new TypeReference<>() {});
+
+        Object existingProperty = claims.get(propertyName);
+        if (existingProperty == null) {
+            Set<Map<String, Object>> roles = new HashSet<>();
+            roles.add(modelMap);
+            claims.put(propertyName, roles);
+        } else if (existingProperty instanceof Set<?> rawSet) {
+            if (rawSet.stream().allMatch(item -> item instanceof Map<?, ?>)) {
+                @SuppressWarnings("unchecked")
+                Set<Map<String, Object>> rolesProperty = (Set<Map<String, Object>>) rawSet;
                 rolesProperty.add(modelMap);
-                claims.put(propertyName, rolesProperty);
             } else {
-                LOGGER.warnf("Incompatible types for property %s. The mapper will not set the roles for client %s",
-                        propertyName, client);
+                LOGGER.warnf("Incompatible types for property %s. Mapper will not set roles for client %s.", propertyName, client);
             }
         } else {
-            // needs to be mutable
-            Set roles = new HashSet();
-            roles.add(modelMap);
-            claims.put(propertyName, roles);
+            LOGGER.warnf("Incompatible types for property %s. Mapper will not set roles for client %s.", propertyName, client);
```
Maybe I am wrong and it happens somewhere else, but the class comment from Stefan says `Adds the users roles to the credential subject`. But looking at that code simply all clientRoles are extracted and returned. Are the roles filtered from the claim at a later point?

Awambeng replied:
@Captain-P-Goldfish thank you for the feedback! I suspect that @wistefan's comment might be outdated, as the code (including the original implementation) deals with client roles rather than user roles. @wistefan  what do you think?

Captain-P-Goldfish replied:
But why should we map the ClientRoles into a credential? I do not see for what purpose it does any good. Mapping the ClientRoles of the client assigned to the user makes sense, but simply adding all ClientRoles to it seems odd.

Awambeng replied:
@wistefan, @tnorimat, @thomasdarimont, @francis-pouatcha  – I’d like to hear your thoughts on this discussion. Do you think mapping all ClientRoles into the credential is appropriate, or should we only include ClientRoles assigned to the user?

tnorimat replied:
@Awambeng If you need to differentiate users for a use case like EUDI wallet, we need to only map client roles assigned to a user. But I am not sure whether we need to differenceate users. It might be good to discuss it in OAuth SIG breakout sessions if you are also not sure.

wistefan replied:
As far as I understand l122 ```userSessionModel.getUser().getClientRoleMappingsStream(clientModel).toList();``` only return ClientRoles that are assigned to the user. Thus, it works as itendended.

Awambeng committed:
Merge branch 'main' into issue-32957

Awambeng committed:
Merge branch 'main' into issue-32957

tnorimat commented:
@Awambeng Hello, thank you the PR. I will review it.

tnorimat commented on services/src/main/java/org/keycloak/protocol/oid4vc/issuance/mappers/OID4VCTargetRoleMapper.java:
```diff
@@ -113,31 +114,42 @@ public void setClaimsForSubject(Map<String, Object> claims,
             } else {
-                LOGGER.warnf("Incompatible types for property %s. The mapper will not set the roles for client %s",
-                        propertyName, client);
+                LOGGER.warnf("Incompatible types for property %s. Mapper will not set roles for client %s.", propertyName, client);
```
Line 149 and Line 152 has the same warning messages. How about having different warning messages?

tnorimat reviewed (commented):
@Awambeng I added some review comments. Could you check them?

Awambeng committed:
chore: address PR comments and improve logging

Signed-off-by: Awambeng Rodrick <awambengrodrick@gmail.com>

Awambeng commented:
@Captain-P-Goldfish @tnorimat  I've reviewed and addressed the comments in this PR. Could you please take another look when you have time? Let me know if anything else needs adjustment. Thanks!

tnorimat reviewed (commented):
LGTM except https://github.com/keycloak/keycloak/pull/37687#discussion_r1988877243

wistefan reviewed (approved):
I checked the implementation of the TargetRoleMapper as discussed today and for my understanding its working as intended, e.g. only including roles that are actually assigned to the user.

LGTM

Captain-P-Goldfish commented:
looks much better now!

tnorimat reviewed (approved):
LGTM

tnorimat commented:
@mposolda our reviewers all approved the PR. Could you check it?

Awambeng commented:
@Captain-P-Goldfish @tnorimat @wistefan @francis-pouatcha  thanks for taking the time to review this PR! I really appreciate your feedback and suggestions.

mposolda reviewed (approved):
@Awambeng @francis-pouatcha @Captain-P-Goldfish @wistefan @tnorimat Thanks everyone for the reviews and feedback on this PR!
